### PR TITLE
[MDS-5813] File Upload for Purpose and Authorization Page

### PR DESCRIPTION
**Pull request by isuru-aot**
## Objective 

[MDS-5813](https://bcmines.atlassian.net/browse/MDS-5813)

_Why are you making this change? Provide a short explanation and/or screenshots_

- Added 3 separate file uploads
- Added summary table which shows the uploaded file.
- TODO `renderCategoryColumn` seems not picking up the file type.

### CODE REVIEW

File: services/common/src/components/projectSummary/AuthorizationsInvolved.tsx [hunk 1]
```diff
@@ -37,16 +37,88 @@ import { fetchPermits } from "@mds/common/redux/actionCreators/permitActionCreat
...
+
+  const tableDocuments =
+    props?.amendment_documents?.map(
+      (doc) => new MineDocument({ ...doc, category: doc.project_summary_document_type_code })
+    ) ?? [];
+
+  const documentColumns = [
+    renderTextColumn("document_name", "File Name"),
+    renderCategoryColumn(
+      "project_summary_document_type_code",
```
Comment (isuru-aot) [inline services/common/src/components/projectSummary/AuthorizationsInvolved.tsx]: This function doesn't seems rendering the file type as expected. I tried `projectSummaryDocumentTypesHash` from the props initially and then tried passing a map as above but nothing worked. And also noticed that its not working in other [document-upload table in the form](https://github.com/bcgov/mds/blob/206487d16e9d80ac1b816d37240e743ff5137bc8/services/minespace-web/src/components/Forms/projects/projectSummary/DocumentUpload.tsx#L107).

Did some debugging and its complaining about `text is undefined` [in the function here](https://github.com/bcgov/mds/blob/206487d16e9d80ac1b816d37240e743ff5137bc8/services/common/src/components/common/CoreTableCommonColumns.tsx#L62)

File: services/minespace-web/src/components/pages/Project/ProjectSummaryPage.tsx [hunk 2]
```diff
@@ -268,6 +268,10 @@ export const ProjectSummaryPage: FC<ProjectSummaryPageProps> = (props) => {
     const newActiveTab = projectFormTabs[currentTabIndex + 1];
     const message = "Successfully saved a draft project description.";
     const values = { ...formValues, status_code: "DFT" };
+    const file_to_upload =
+      values.authorizations.AIR_EMISSIONS_DISCHARGE_PERMIT.AMENDMENT[0].amendment_documents;
+    const filtered = file_to_upload.filter((doc) => !doc.mine_document_guid);
+    values.authorizations.AIR_EMISSIONS_DISCHARGE_PERMIT.AMENDMENT[0].amendment_documents = filtered;
```
Comment (matbusby-fw) [inline services/minespace-web/src/components/pages/Project/ProjectSummaryPage.tsx:276]: I don't think mutating the form state like this will work since you're reassigning a form value directly.  For that you can use the redux-form `change` function.  You'd need to add it to the mapDispatchToProps for this file and then this line could like more like this:

```
props.change(
      FORM.ADD_EDIT_PROJECT_SUMMARY,
      values.authorizations.AIR_EMISSIONS_DISCHARGE_PERMIT.AMENDMENT[0].amendment_documents,
      filtered
    );
```

File: services/common/src/components/projectSummary/AuthorizationsInvolved.tsx (hunk 1 above)
Comment (matbusby-fw) [reply services/common/src/components/projectSummary/AuthorizationsInvolved.tsx]: Took a look.  The document records in the table look like this:

```
{
    "mine_document_guid": "c2b1b949-26e2-45a2-9f20-160851a174fd",
    "mine_guid": "34af5a68-e1f6-47c8-9aa8-ec4b49180e1d",
    "document_manager_guid": "3730a84c-4d76-4238-b067-fd644e543914",
    "document_name": "test1.pdf",
    "create_user": "idir\\matbusby-ey@bceid",
    "upload_date": "2024-05-06 17:37:15.986560+00:00",
    "update_timestamp": null,
    "category": "MAP",
    "is_archived": false,
    "archived_by": null,
    "archived_date": null,
    "is_latest_version": true,
    "entity_title": "",
    "key": "c2b1b949-26e2-45a2-9f20-160851a174fd",
    "file_type": ".pdf",
    "number_prev_versions": 0,
    "versions": [],
    "allowed_actions": [
        "Open in document viewer",
        "Download file",
        "Replace file",
        "Archive file",
        "Delete"
    ]
}
```

The first property you pass to `renderCategoryColumn` will be the data index which has to reference the data within the record object.  So, since the record simply has a `category` field rather than a `project_summary_document_type_code`, changing that first prop to `category` should do the trick.

File: services/common/src/components/projectSummary/AuthorizationSupportDocumentUpload.tsx [hunk 3]
```diff
@@ -0,0 +1,129 @@
+import React, { FC } from "react";
+import { Field, getFormValues } from "redux-form";
+import { DOCUMENT, EXCEL, IMAGE, SPATIAL } from "@mds/common/constants/fileTypes";
+import { MINE_REPORT_DOCUMENT, PROJECT_SUMMARY_DOCUMENTS } from "@mds/common/constants/API";
+import RenderFileUpload from "../forms/RenderFileUpload";
+import { required, requiredNewFiles } from "@mds/common/redux/utils/Validate";
+import { IProjectSummaryDocument, IMineDocument } from "../..";
+import { useSelector } from "react-redux";
+import { FORM } from "../..";
```
Comment (matbusby-fw) [inline services/common/src/components/projectSummary/AuthorizationSupportDocumentUpload.tsx]: A bit nitpicky, but there are a few unused imports that could be removed here.

File: services/common/src/components/projectSummary/AuthorizationSupportDocumentUpload.tsx [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    const newDocuments = documents.filter(
+      (file) => fileToRemove.serverId !== file.document_manager_guid
+    );
+    newDocuments.forEach((newDoc) => {
+      updateAmendmentDocuments(newDoc);
+    });
+  };
+
+  const handleFileLoad = (
+    document_name,
```
Comment (matbusby-fw) [inline services/common/src/components/projectSummary/AuthorizationSupportDocumentUpload.tsx]: Small thing, but it would be nice to have these props typed.

ie:

```
    document_name: string,
    document_manager_guid: string,
    project_summary_document_type_code: string
```

File: services/common/src/components/projectSummary/AuthorizationsInvolved.tsx [hunk 5]
```diff
@@ -37,16 +37,88 @@ import { fetchPermits } from "@mds/common/redux/actionCreators/permitActionCreat
...
 import { FORM } from "@mds/common/constants/forms";
 import RenderHiddenField from "../forms/RenderHiddenField";
+import AuthorizationSupportDocumentUpload from "./AuthorizationSupportDocumentUpload";
+import { IProjectSummaryDocument } from "@mds/common";
+import {
+  renderTextColumn,
+  renderDateColumn,
+} from "@mds/common/components/common/CoreTableCommonColumns";
+import DocumentTable from "@mds/common/components/documents/DocumentTable";
+import {
```
Comment (matbusby-fw) [inline services/common/src/components/projectSummary/AuthorizationsInvolved.tsx]: I think this import isn't being used.  It can probably go.

File: services/common/src/components/projectSummary/AuthorizationsInvolved.tsx [hunk 6]
```diff
@@ -454,8 +628,12 @@ export const AuthorizationsInvolved = () => {
                                   />
```
Comment (matbusby-fw) [inline services/common/src/components/projectSummary/AuthorizationsInvolved.tsx:609]: This is from a previous PR it seems, but just noticed a typo on line 622 `Authorixation` should be `Authorization` and `FrountCounter BC` should be `FrontCounter BC`

File: services/common/src/components/projectSummary/AuthorizationsInvolved.tsx [hunk 7]
```diff
@@ -37,16 +37,88 @@ import { fetchPermits } from "@mds/common/redux/actionCreators/permitActionCreat
...
+  status_code: string;
+  submission_date: string;
+  project_summary_description: string;
+  project_guid: string;
+  documents: IProjectSummaryDocument[];
+}
 
-const RenderEMAPermitCommonSections = ({ isAmendment }) => {
-  const purposeLabel = isAmendment
+interface DocumentUploadProps {
```
Comment (matbusby-fw) [inline services/common/src/components/projectSummary/AuthorizationsInvolved.tsx]: It looks like this isn't being used anywhere.  Is it still needed?

File: services/core-api/app/api/projects/project_summary/models/project_summary.py [hunk 8]
```diff
@@ -315,6 +329,20 @@ def create_or_update_authorization(self, authorization):
                     ams_tracking_number=authorization.get('ams_tracking_number'),
                     ams_outcome=authorization.get('ams_outcome')
                 )
+# Check only for new files
```
Comment (matbusby-fw) [inline services/core-api/app/api/projects/project_summary/models/project_summary.py:333]: Another tiny nitpick, but would be nice to have this indented to the level of the fuction.

File: services/core-api/app/api/projects/project_summary/models/project_summary_authorization.py [hunk 9]
```diff
@@ -199,11 +209,12 @@ def create(cls,
...
             ams_tracking_number=ams_tracking_number,
             ams_outcome=ams_outcome,
-            ams_status_code=ams_status_code)
+            ams_status_code=ams_status_code,
+            amendment_documents=amendment_documents)
         if add_to_session:
             new_authorization.save(commit=False)
         return new_authorization
-
+# TODO fileupload only when uploading
```
Comment (matbusby-fw) [inline services/core-api/app/api/projects/project_summary/models/project_summary_authorization.py]: I'm not sure I understand the purpose of this TODO.

If it's referencing work that is outside the scope of this ticket, can we make it a bit more descriptive?

File: services/common/src/components/projectSummary/AuthorizationSupportDocumentUpload.tsx (hunk 3 above)
Comment (isuru-aot) [reply services/common/src/components/projectSummary/AuthorizationSupportDocumentUpload.tsx]: > A bit nitpicky, but there are a few unused imports that could be removed here.

Updated accordingly.

File: services/common/src/components/projectSummary/AuthorizationSupportDocumentUpload.tsx (hunk 4 above)
Comment (isuru-aot) [reply services/common/src/components/projectSummary/AuthorizationSupportDocumentUpload.tsx]: > Small thing, but it would be nice to have these props typed.
> 
> ie:
> 
> ```
>     document_name: string,
>     document_manager_guid: string,
>     project_summary_document_type_code: string
> ```

Updated accordingly.

File: services/common/src/components/projectSummary/AuthorizationsInvolved.tsx (hunk 5 above)
Comment (isuru-aot) [reply services/common/src/components/projectSummary/AuthorizationsInvolved.tsx]: yes. Removed.

File: services/common/src/components/projectSummary/AuthorizationsInvolved.tsx (hunk 7 above)
Comment (isuru-aot) [reply services/common/src/components/projectSummary/AuthorizationsInvolved.tsx]: Nope, Removed.

File: services/common/src/components/projectSummary/AuthorizationsInvolved.tsx (hunk 1 above)
Comment (isuru-aot) [reply services/common/src/components/projectSummary/AuthorizationsInvolved.tsx]: oh, I was playing with the map. Thanks. Updated accordingly and now it works.

File: services/common/src/components/projectSummary/AuthorizationsInvolved.tsx (hunk 6 above)
Comment (isuru-aot) [reply services/common/src/components/projectSummary/AuthorizationsInvolved.tsx:609]: Corrected.

File: services/minespace-web/src/components/pages/Project/ProjectSummaryPage.tsx (hunk 2 above)
Comment (isuru-aot) [reply services/minespace-web/src/components/pages/Project/ProjectSummaryPage.tsx:276]: Thank you. Updated as suggested.

File: services/core-api/app/api/projects/project_summary/models/project_summary_authorization.py (hunk 9 above)
Comment (matbusby-fw) [reply services/core-api/app/api/projects/project_summary/models/project_summary_authorization.py]: Not sure if you saw this comment @isuru-aot.  Would be good to have more clarity on what is left TODO here.

Comment (isuru-aot) [reply services/core-api/app/api/projects/project_summary/models/project_summary_authorization.py]: Yeah, I missed this. Actually this was for my reference and forgot to remove it. The logic for that was moved to project_summary.py

I just removed this.

**Review by matbusby-fw — DISMISSED** [review]
Another tough one, but good work Isuru!

File: services/common/src/components/forms/RenderFileUpload.tsx [hunk 10]
```diff
@@ -440,6 +442,13 @@ export const FileUpload = (props: FileUploadProps) => {
   const fileValidateTypeLabelExpectedTypesMap = invert(props.acceptedFileTypesMap);
   const acceptedFileTypes = uniq(Object.values(props.acceptedFileTypesMap));
 
+  const getLabel = (props) => {
+    if (props.labelHref)
+      return <a href={props.labelHref} target="_blank">{props.label}</a>
```
Comment (henryoforeh-dev) [inline services/common/src/components/forms/RenderFileUpload.tsx]: Potential security risk here, needs a little bit of enhancement with the rel="noopener noreferrer" attribute.

**Review by matbusby-fw — APPROVED** [review]

File: services/common/src/components/forms/RenderFileUpload.tsx (hunk 10 above)
Comment (isuru-aot) [reply services/common/src/components/forms/RenderFileUpload.tsx]: Thanks. This updated accordingly @henryoforeh-aot

**Review by henryoforeh-dev — APPROVED** [review]
